crates/providers/kv-vault/src/lib.rs

```diff
@@ -1,132 +1,123 @@
+use wasmcloud_provider_sdk::Context;
 
-    eprintln!("KvVault provider exiting");
-    Ok(())
-}
+pub(crate) mod client;
+pub(crate) mod config;
+pub(crate) mod error;
+
+use crate::client::Client;
+use crate::config::Config;
+use crate::error::VaultError;
+
+/// Token to indicate string data was passed during set
+pub const STRING_VALUE_MARKER: &str = "string_data___";
 
-/// Redis keyValue provider implementation.
-#[derive(Default, Clone, Provider)]
-#[services(KeyValue)]
-struct KvVaultProvider {
+wasmcloud_provider_wit_bindgen::generate!({
+    impl_struct: KvVaultProvider,
+    contract: "wasmcloud:keyvalue",
+    wit_bindgen_cfg: "provider-kv-vault"
+});
+
+/// Redis KV provider implementation which utilizes [Hashicorp Vault](https://developer.hashicorp.com/vault/docs)
+#[derive(Default, Clone)]
+pub struct KvVaultProvider {
     // store redis connections per actor
     actors: std::sync::Arc<RwLock<HashMap<String, RwLock<Client>>>>,
 }
-/// use default implementations of provider message handlers
-impl ProviderDispatch for KvVaultProvider {}
 
-/// Handle provider control commands
-/// put_link (new actor link command), del_link (remove link command), and shutdown
+impl KvVaultProvider {
+    /// Retrieve a client for a given context (determined by actor_id)
+    async fn get_client(&self, ctx: &Context) -> ProviderInvocationResult<Client> {
+        // get the
+        let actor_id = ctx.actor.as_ref().ok_or_else(|| {
+            ProviderInvocationError::Provider("invalid parameter: no actor in request".into())
+        })?;
+        // Clone the existing client for the given actor from the internal hash map
+        let client = self
+            .actors
+            .read()
+            .await
+            .get(actor_id)
+            .ok_or_else(|| {
+                ProviderInvocationError::Provider(format!(
+                    "invalid parameter: actor [{actor_id}] not linked"
+                ))
+            })?
+            .read()
+            .await
+            .clone();
+        Ok(client)
+    }
+}
+
+/// Handle provider control commands, the minimum required of any provider on
+/// a wasmcloud lattice
 #[async_trait]
-impl ProviderHandler for KvVaultProvider {
+impl WasmcloudCapabilityProvider for KvVaultProvider {
     /// Provider should perform any operations needed for a new link,
     /// including setting up per-actor resources, and checking authorization.
     /// If the link is allowed, return true, otherwise return false to deny the link.
     #[instrument(level = "debug", skip(self, ld), fields(actor_id = %ld.actor_id))]
-    async fn put_link(&self, ld: &LinkDefinition) -> RpcResult<bool> {
-        let config = Config::from_values(&ld.values)?;
-        let client = Client::new(config).map_err(to_rpc_err)?;
+    async fn put_link(&self, ld: &LinkDefinition) -> bool {
+        let config = if let Ok(config) =
+            Config::from_values(&HashMap::from_iter(ld.values.clone().into_iter()))
+        {
```

Comment:
nit: This seems like an excessive conversion? Is `ld.values` not already a hashmap?

Reply:
`values` is a `WitMap<String>` (a list of tuples) 😭